File: polyglot/Decorator/decorator.py
# Decorator Design Pattern

# The implementation of the decorator pattern in python does not follow
# the pattern encountered in other languages, due to python support for
# multiple inheritance


class Pizza(object):
    @property
    def description(self):
        raise NotImplementedError("Pizza::description is not implemented")
    @property
    def cost(self):
        raise NotImplementedError("Pizza::cost is not implemented")


class PlainPizza(Pizza):
    @property
    def description(self): return "Plain pizza"
    @property
    def cost(self): return 3.0

class WithExtraCheese(object):
    @property
    def description(self): return super().description + " + extra cheese" 
    @property
    def cost(self): return super().cost + 0.5

class WithExtraOlives(object):
    @property
    def description(self): return super().description + " + extra olives" 
    @property
    def cost(self): return super().cost + 0.7

class WithExtraAnchovies(object):
    @property
    def description(self): return super().description + " + extra anchovies" 
    @property
    def cost(self): return super().cost + 0.8

class NicePizza(WithExtraCheese, PlainPizza)    : pass
class SuperPizza(WithExtraOlives, NicePizza)    : pass
class UltraPizza(WithExtraAnchovies, SuperPizza): pass

plainPizza  = PlainPizza()
nicePizza   = NicePizza()
superPizza  = SuperPizza()
ultraPizza  = UltraPizza()


print(  "Plain Pizza: "     +
        "\tCost: "          + str(plainPizza.cost) +
        "\tDescription: "   + plainPizza.description)

print(  "Nice Pizza: "      +
        "\tCost: "          + str(nicePizza.cost) +
        "\tDescription: "   + nicePizza.description)

print(  "Super Pizza: "     +
        "\tCost: "          + str(superPizza.cost) +
        "\tDescription: "   + superPizza.description)

print(  "Ultra Pizza: "     +
        "\tCost: "          + str(ultraPizza.cost) +
        "\tDescription: "   + ultraPizza.description)

File: polyglot/Decorator/decorator.js
// Decorator Design Pattern


// interface or abstract class of components
////////////////////////////////////////////////////////////////////////////////
//                                 Pizza                                      //
////////////////////////////////////////////////////////////////////////////////
function Pizza(){}

Pizza.prototype.getDescription = function(){
  throw "Pizza::getDescription is not implemented";
}

Pizza.prototype.getCost = function(){
  throw "Pizza::getCost is not implemented";
}

////////////////////////////////////////////////////////////////////////////////
//                               Plain Pizza                                  //
////////////////////////////////////////////////////////////////////////////////

// concrete component
function PlainPizza(){ Pizza.call(this); }

PlainPizza.prototype = Object.create(Pizza.prototype);
PlainPizza.prototype.getDescription = function(){ return "Plain pizza"; }
PlainPizza.prototype.getCost = function(){ return 3.0; }

// root abstract class of decorators implementing component interface
// a decorator object is therefore a component and will hold a 
// concrete component as data member. Default implementation
// achieved simply by forwarding 


////////////////////////////////////////////////////////////////////////////////
//                                PizzaDecorator                              //
////////////////////////////////////////////////////////////////////////////////

function PizzaDecorator(pizza) {
  Pizza.call(this);
  this.pizza = pizza;
}
PizzaDecorator.prototype = Object.create(Pizza.prototype);
Pizza.prototype.getDescription = function(){ return this.pizza.getDescription(); }
Pizza.prototype.getCost = function(){ return this.pizza.getCost(); }

////////////////////////////////////////////////////////////////////////////////
//                                WithExtraCheese                             //
////////////////////////////////////////////////////////////////////////////////

// first concrete decorator
function WithExtraCheese(pizza){ PizzaDecorator.call(this, pizza); }
WithExtraCheese.prototype = Object.create(PizzaDecorator.prototype);
WithExtraCheese.prototype.getDescription = function(){
  return this.pizza.getDescription() + " + extra cheese";
}
WithExtraCheese.prototype.getCost = function(){
    return this.pizza.getCost() + 0.5;
}

////////////////////////////////////////////////////////////////////////////////
//                                WithExtraOlives                             //
////////////////////////////////////////////////////////////////////////////////

// second concrete decorator
function WithExtraOlives(pizza){ PizzaDecorator.call(this, pizza); }
WithExtraOlives.prototype = Object.create(PizzaDecorator.prototype);
WithExtraOlives.prototype.getDescription = function(){
  return this.pizza.getDescription() + " + extra olives";
}
WithExtraOlives.prototype.getCost = function(){
    return this.pizza.getCost() + 0.7;
}

////////////////////////////////////////////////////////////////////////////////
//                              WithExtraAnchovies                            //
////////////////////////////////////////////////////////////////////////////////

// third concrete decorator
function WithExtraAnchovies(pizza){ PizzaDecorator.call(this, pizza); }
WithExtraAnchovies.prototype = Object.create(PizzaDecorator.prototype);
WithExtraAnchovies.prototype.getDescription = function(){
  return this.pizza.getDescription() + " + extra anchovies";
}
WithExtraAnchovies.prototype.getCost = function(){
    return this.pizza.getCost() + 0.8;
}


////////////////////////////////////////////////////////////////////////////////
//                                  Main                                      //
////////////////////////////////////////////////////////////////////////////////

plainPizza  = new PlainPizza();

nicePizza   = new WithExtraCheese(
              new PlainPizza()); 

superPizza  = new WithExtraOlives( 
              new WithExtraCheese(
              new PlainPizza())); 

ultraPizza  = new WithExtraAnchovies(
              new WithExtraOlives( 
              new WithExtraCheese(
              new PlainPizza()))); 

print(  "Plain Pizza: "   +
        "\tCost: "        + plainPizza.getCost().toFixed(1) // formatting
      + "\tDescription: " + plainPizza.getDescription()); 


print(  "Nice Pizza: "   +
        "\tCost: "        + nicePizza.getCost().toFixed(1)
      + "\tDescription: " + nicePizza.getDescription()); 

print(  "Super Pizza: "   +
        "\tCost: "        + superPizza.getCost().toFixed(1)
      + "\tDescription: " + superPizza.getDescription()); 

print(  "Ultra Pizza: "   +
        "\tCost: "        + ultraPizza.getCost().toFixed(1)
      + "\tDescription: " + ultraPizza.getDescription()); 


